### PR TITLE
Handle disabled variables v2

### PR DESCRIPTION
**Pull request by codenirvana**
* Handle `disabled` property in:  
  * `VariableScope~has` : returns false if variable is disabled
  * `VariableScope~get` : returns undefined if variable is disabled or return last enabled
  * `VariableScope~unset` : remove ~last~ all enabled or noop if disabled
  * `VariableScope~set` : if disabled, create new enabled variable
* Handle `disabled` property in `Varaible~update`
* Resolve last enabled variable

Closes #778

### CODE REVIEW

File: lib/collection/variable-list.js [hunk 1]
```diff
@@ -131,6 +131,41 @@ _.assign(VariableList.prototype, /** @lends VariableList.prototype */ {
...
+     */
+    normalizeReference: function (variable) {
+        var i,
+            newVar,
+            index = this._postman_listIndexKey; // `key` for Variable
+
+        // traverse the members list in reverse direction in order to find the last enabled
+        for (i = this.members.length - 1; i >= 0; i--) {
+            newVar = this.members[i];
+            if (newVar[index] === variable[index] && !newVar.disabled) {
```
Comment (kunagpal) [inline lib/collection/variable-list.js]: Assign `variable[index]` to a variable and reuse it here.

File: lib/collection/variable-scope.js [hunk 2]
```diff
@@ -184,7 +194,15 @@ _.assign(VariableScope.prototype, /** @lends VariableScope.prototype */ {
      * @returns {Boolean}
      */
     has: function (variableName) {
-        return this.values.has(variableName);
+        var variable = this.values.one(variableName);
+
```
Comment (kunagpal) [inline lib/collection/variable-scope.js]: Extra blank line.

File: lib/collection/variable-scope.js [hunk 3]
```diff
@@ -244,7 +277,45 @@ _.assign(VariableScope.prototype, /** @lends VariableScope.prototype */ {
      * @param {String} key
      */
     unset: function (key) {
-        this.values.remove(key);
+        var i,
+            member,
+            newReference,
+            variable = this.values.one(key);
+
+        // bail out if variable doesn't exists
```
Comment (kunagpal) [inline lib/collection/variable-scope.js]: exist

File: lib/collection/variable-scope.js [hunk 4]
```diff
@@ -244,7 +277,45 @@ _.assign(VariableScope.prototype, /** @lends VariableScope.prototype */ {
...
+
+        // bail out if variable is disabled
+        // @note disabled variables are immutable
+        if (variable.disabled) {
+            return;
+        }
+
+        // traverse the members list in reverse direction in order to find the
+        // last enabled variable.
+        for (i = this.values.members.length - 1; i >= 0; i--) {
```
Comment (kunagpal) [inline lib/collection/variable-scope.js]: Store and reuse `this.values.members`

File: test/unit/variable-scope.test.js [hunk 5]
```diff
@@ -341,11 +382,27 @@ describe('VariableScope', function () {
                 expect(scope.get('var-1')).to.equal('new-var-1-value');
             });
 
+            it('should correctly update the last enabled item in multi value list', function () {
+                scope.set('var-2', 'new-var-2-value');
+                expect(scope.get('var-2')).to.equal('new-var-2-value');
```
Comment (kunagpal) [inline test/unit/variable-scope.test.js]: Also check for the right index being updated.

File: test/unit/variable-scope.test.js [hunk 6]
```diff
@@ -341,11 +382,27 @@ describe('VariableScope', function () {
...
+            });
+
+            it('should handle disabled variable correctly', function () {
```
Comment (kunagpal) [inline test/unit/variable-scope.test.js]: Tests are dependent on each other, this shouldn't be the case.

File: lib/collection/variable-scope.js [hunk 7]
```diff
@@ -244,7 +262,34 @@ _.assign(VariableScope.prototype, /** @lends VariableScope.prototype */ {
...
-        this.values.remove(key);
+        var list = this.values;
+
+        // early bail out if there's no variable with the given name
+        if (!list.one(key)) {
+            return;
+        }
+
+        // remove all enabled members from the list
+        _.remove(list.members, function (variable) {
```
Comment (shamasis) [inline lib/collection/variable-scope.js]: send it to underlying remove function

File: lib/collection/variable-scope.js [hunk 8]
```diff
@@ -244,7 +260,28 @@ _.assign(VariableScope.prototype, /** @lends VariableScope.prototype */ {
      * @param {String} key
      */
     unset: function (key) {
-        this.values.remove(key);
+        var disabledVariable;
```
Comment (kunagpal) [inline lib/collection/variable-scope.js]: This should be called `lastDisabledVariable` to express intent better.

File: test/unit/variable.test.js [hunk 9]
```diff
@@ -38,6 +38,18 @@ describe('Variable', function () {
         expect(v.system).to.be.false;
     });
 
+    it('should update the disabled property of a variable', function () {
+        var v = new Variable();
+        v.update({ disabled: true });
+        expect(v.disabled).to.be.true;
+
+        v = new Variable({
```
Comment (kunagpal) [inline test/unit/variable.test.js]: Assign this to a separate `variable`

**Pull request merged**